### PR TITLE
FIX: warn when covariance matrix is singular in LinearExplainer

### LINKED ISSUE #4469: BUG: LinearExplainer produces unreliable SHAP values when n_samples <= n_features
### Issue Description

In `shap/explainers/_linear.py`, when `feature_perturbation="correlation_dependent"` is used with a background dataset where `n_samples <= n_features`, the covariance matrix is singular (rank-deficient). SHAP silently produces numerically unreliable attribution values with no warning to the user.

### Minimal Reproducible Example

```python
import numpy as np
from sklearn.linear_model import LinearRegression
import shap

rng = np.random.default_rng(42)
n_features = 10

# 8 samples, 10 features — fewer samples than features
X_train = rng.normal(size=(8, n_features))
y_train = X_train @ np.arange(1, n_features + 1) + rng.normal(size=8)
X_test  = rng.normal(size=(50, n_features))

model = LinearRegression().fit(X_train, y_train)

explainer = shap.LinearExplainer(
    model,
    X_train,
    feature_perturbation="correlation_dependent"
)
shap_values = explainer.shap_values(X_test)
print("SHAP values computed silently with no warning.")
print("Covariance matrix condition number:",
      np.linalg.cond(np.cov(X_train, rowvar=False)))
```

### Traceback

```shell
No exception is raised (related to calculation of SHAP values). SHAP completes silently and returns values that appear plausible but are mathematically unreliable.
```

### Evidence of Incorrect Behavior
Comparing correlation_dependent SHAP against interventional SHAP (ground truth for independent features) across a well-sampled (n=500) and under-sampled (n=8) background dataset:

| Metric                    | Good (n=500) | Bad (n=8)                 |
|---------------------------|--------------|---------------------------|
| Mean per-feature error    | 0.3788         | 3.2049 (8.4x worse)           |
| Sign agreement            | 97.6%        | 71.2%                     |

Image

The code for this is:
```
import numpy as np
from sklearn.linear_model import LinearRegression
import shap

rng = np.random.default_rng(42)
n_features = 10
true_coef = np.arange(1, n_features + 1, dtype=float)

X_test = rng.normal(size=(50, n_features))

# --- Good case ---
X_train_good = rng.normal(size=(500, n_features))
y_train_good = X_train_good @ true_coef + rng.normal(size=500)
model_good = LinearRegression().fit(X_train_good, y_train_good)

explainer_good = shap.LinearExplainer(
    model_good, X_train_good,
    feature_perturbation="correlation_dependent"
)
shap_good = explainer_good.shap_values(X_test)

# interventional SHAP for independent features = coef * (x - mean)
# this is the "ground truth" attribution for uncorrelated features
interventional_good = (X_test - X_train_good.mean(0)) * model_good.coef_

# --- Bad case ---
X_train_bad = rng.normal(size=(8, n_features))
y_train_bad = X_train_bad @ true_coef + rng.normal(size=8)
model_bad = LinearRegression().fit(X_train_bad, y_train_bad)

import warnings
with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    explainer_bad = shap.LinearExplainer(
        model_bad, X_train_bad,
        feature_perturbation="correlation_dependent"
    )
shap_bad = explainer_bad.shap_values(X_test)

interventional_bad = (X_test - X_train_bad.mean(0)) * model_bad.coef_

# --- Compare per-feature attribution error vs interventional ground truth ---
# for independent features, correlation_dependent should ≈ interventional
error_good = np.abs(shap_good - interventional_good).mean(axis=0)
error_bad  = np.abs(shap_bad  - interventional_bad ).mean(axis=0)

print("Per-feature mean absolute deviation from interventional SHAP:")
print(f"{'Feature':<10} {'Good (n>>m)':>15} {'Bad (n<=m)':>15}")
print("-" * 40)
for i in range(n_features):
    print(f"feature_{i:<3} {error_good[i]:>15.4f} {error_bad[i]:>15.4f}")

print()
print(f"{'Mean across all features':<10} {error_good.mean():>15.4f} {error_bad.mean():>15.4f}")
print()

# --- Show sign flips specifically ---
print("Sign agreement between correlation_dependent and interventional SHAP:")
sign_good = np.sign(shap_good) == np.sign(interventional_good)
sign_bad  = np.sign(shap_bad)  == np.sign(interventional_bad)
print(f"Good case sign agreement: {sign_good.mean()*100:.1f}%")
print(f"Bad case sign agreement : {sign_bad.mean()*100:.1f}%")
print()
print("A sign flip means a feature that truly DECREASED the prediction")
print("is being reported as INCREASING it, or vice versa.")
```

### Expected Behavior
When` n_samples <= n_features`, SHAP should raise a UserWarning before computing transforms, informing the user that the covariance matrix is singular and the results may be unreliable.
Expected warning:
`UserWarning: The number of samples (8) is less than or equal to the number of features (10). This will produce a singular covariance matrix and may result in unreliable SHAP values when using feature_perturbation='correlation_dependent'.`

### Current Behavior
No warning is raised. SHAP silently applies a small regularization `(1e-6)` to the covariance matrix and continues, masking the problem rather than surfacing it.

### Bug report checklist

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [x] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [x] I'd be interested in making a PR to fix this bug

### Installed Versions

`0.51.0`

AI Usage: The concept of singular matrices affecting the calculation of SHAP values and the performance of LinearExplainer was explained to me by Claude. The fix was to raise a warning notifying the user about the quality of values produced. It is well understood by me.

### PR DESCRIPTION
**Pull request by BaibhavKundu2005**
## What does this PR do?
Adds a `UserWarning` in `LinearExplainer` when the background dataset has fewer samples than features, which produces a singular covariance matrix and unreliable SHAP attributions.

## What problem does it solve?
Closes #4469 

When `n_samples <= n_features`, the covariance matrix is rank-deficient. SHAP previously applied silent regularization (1e-6) and continued, producing values that look plausible but are mathematically unreliable.

## Evidence
Tested with n=500 (good) vs n=8 (bad) background datasets, 10 features:

| Metric                    | Good (n=500) | Bad (n=8)                 |
|---------------------------|--------------|---------------------------|
| Mean per-feature error    | 0.3788         | 3.2049 (8.4x worse)           |
| Sign agreement            | 97.6%        | 71.2%                     |

Sign agreement of 71.2% means 1 in 4 features attributed in the wrong direction on a beeswarm plot, with no indication anything is wrong.

## Fix
Guard the `np.cov` call with a sample count check and raise a `UserWarning` before computing transforms. The regularization already present is kept as a safety net, but users are now informed before results become unreliable.

## Checklist
- [x] I have read CONTRIBUTING.md
- [x] Branch is up to date with master
- [x] Pre-commit hooks pass
- [x] Unit test added covering the singular covariance case

### CODE REVIEW

File: tests/explainers/test_linear.py [hunk 1]
```diff
@@ -262,3 +262,34 @@ def test_interventional_multi_regression():
...
+
+
+def test_linear_explainer_warns_singular_covariance():
+    """LinearExplainer should warn when n_samples <= n_features."""
+    import warnings
+
+    import numpy as np
+    from sklearn.linear_model import LinearRegression
+
+    import shap
```
Comment (CloseChoice) [inline tests/explainers/test_linear.py]: isn't shap imported at the top already?

**Review by CloseChoice — APPROVED** [review]
Looks good, thanks for the PR. Let's wait for the pipelines to pass.

**Comment by BaibhavKundu2005** [comment]
@CloseChoice Thanks for the review and the approval. I removed the redundant imports and resolved the merge conflict with master that popped up. Everything is up to date and ready for the CI pipelines whenever you have a chance to approve the workflow run.

File: tests/explainers/test_linear.py (hunk 1 above)
Comment (BaibhavKundu2005) [reply tests/explainers/test_linear.py]: Apologies. I have removed it now.

**Comment by CloseChoice** [comment]
@BaibhavKundu2005 just did, will come back and merge once pipelines are green. No action needed on your end if everything works fine

**Comment by BaibhavKundu2005** [comment]
@CloseChoice The mypy issue with `data.shape` being flagged as a tuple has been fixed using `np.shape()`. Pinging you one last time to approve the workflow run for this new commit whenever you have a second.

**Comment by BaibhavKundu2005** [comment]
@CloseChoice Hi, all the checks have passed. Let me know if anything else is needed before merging!
